### PR TITLE
Bug 1386336 - Generate robots.txt at checksetup.pl from a template

### PR DESCRIPTION
**Pull request by umohm1**

### CODE REVIEW

**Comment by floatingatoll** [comment]
Reviewers, when this patch is submitted for review, please inspect the HTTP response headers for `curl -v .../robots.txt` and make sure they look correct. (robots.txt can't be immutable, for instance, so it shouldn't have max-age: 1 year.)

File: Bugzilla/Install/Filesystem.pm [hunk 1]
```diff
@@ -405,8 +404,17 @@ sub FILESYSTEM {
...
                                        overwrite => 1,
                                        contents  => $yui3_all_css },
+        "robots.txt"              => { perms     => $CGI_READ,
+                                       overwrite => 1,
+                                       contents  => \&robots_txt},
     );
 
+    sub robots_txt {
+    my $output = '';
+    Bugzilla->template->process("global/robots.txt.html", {}, \$output);
```
Comment (dklawren) [inline Bugzilla/Install/Filesystem.pm]: In the git move, you put the robots.txt into template/en/default/robots.txt.tmpl. You need to move it to .../global/robots.txt.tmpl.

File: extensions/SiteMapIndex/Extension.pm [hunk 2]
```diff
@@ -90,9 +90,6 @@ sub install_before_final_checks {
     }
 
     return if (correct_urlbase() ne 'https://bugzilla.mozilla.org/');
-
-
-    $self->_fix_robots_txt();
```
Comment (dklawren) [inline extensions/SiteMapIndex/Extension.pm]: 1) You need to remove the _fix_robots_txt() function as well.
2) You need to create an extensions/SiteMapIndex/template/en/default/hook/global/robots-end.txt.tmpl file that adds in the stuff that the _fix_robots_txt() function would normally do.

File: Bugzilla/Install/Filesystem.pm [hunk 3]
```diff
@@ -948,6 +950,13 @@ sub _check_web_server_group {
     return $group_id;
 }
 
+sub robots_txt {
+    my $output = '';
+    Bugzilla->template->process("global/robots.txt.tmpl", {SITEMAP_URL => correct_urlbase().SITEMAP_URL}, \$output) or die Bugzilla->template->error;
```
Comment (dylanwh) [inline Bugzilla/Install/Filesystem.pm]: I steered you wrong here. Here's a revised bit of code that add a before_robots_txt hook:
```perl
my %vars;
Bugzilla::Hook::process("before_robots_txt", { vars => \%vars });
Bugzilla->template->process("robots.txt.tmpl", \%vars, \$output)
    or die Bugzilla->template->error;
```

File: extensions/SiteMapIndex/Extension.pm [hunk 4]
```diff
@@ -127,16 +124,6 @@ EOT
...
-    my ($self) = @_;
-    my $cgi_path = bz_locations()->{'cgi_path'};
-    my $robots_file = "$cgi_path/robots.txt";
-    my $current_fh = new IO::File("$cgi_path/robots.txt", 'r');
-    if (!$current_fh) {
-        warn "$robots_file: $!";
-        return;
-    }
-
     my $current_contents;
```
Comment (dylanwh) [inline extensions/SiteMapIndex/Extension.pm]: delete the rest of the lines until _PACKAGE__->NAME; as well (_PACKAGE__->NAME should be the last line)

Before _PACKAGE__->NAME; add the following code (which is the before_robots_txt hook, which puts a single template variable, SITEMAP_URL)

```perl
sub before_robots_txt {
    my ($self, $args) = @_;
    $args->{vars}{SITEMAP_URL} = correct_urlbase() . SITEMAP_URL;
}
```

File: extensions/SiteMapIndex/template/en/default/hook/global/robots-end.txt.tmpl [hunk 5]
```diff
@@ -0,0 +1,20 @@
+User-agent: *
```
Comment (dylanwh) [inline extensions/SiteMapIndex/template/en/default/hook/global/robots-end.txt.tmpl]: Rename this file to xtensions/SiteMapIndex/template/en/default/hook/robots-end.txt.tmpl

File: template/en/default/robots.txt.tmpl [hunk 6]
```diff
@@ -18,3 +18,5 @@ Allow: /describekeywords.cgi
 
 Allow: /data/SiteMapIndex/sitemap*.xml.gz
```
Comment (dylanwh) [inline template/en/default/robots.txt.tmpl]: delete line 19 and line 20

File: Bugzilla/Install/Filesystem.pm [hunk 7]
```diff
@@ -948,6 +950,12 @@ sub _check_web_server_group {
     return $group_id;
 }
 
+my %vars;
```
Comment (dylanwh) [inline Bugzilla/Install/Filesystem.pm]: before this line, add sub robots_txt {

File: Bugzilla/Install/Filesystem.pm [hunk 8]
```diff
@@ -948,6 +950,12 @@ sub _check_web_server_group {
...
+Bugzilla::Hook::process("before_robots_txt", { vars => \%vars });
+Bugzilla->template->process("robots.txt.tmpl", \%vars, \$output)
+    or die Bugzilla->template->error;
+
+
```
Comment (dylanwh) [inline Bugzilla/Install/Filesystem.pm]: add a } here

**Pull request merged**